torch/_inductor/autotune_process.py

```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
+    ) -> Callable[[], None]:
+        mod = PyCodeCache.load_by_key_path(self.module_cache_key, self.module_path)
+        if DEBUG:
+            print(
+                f"benchmark module key: {self.module_cache_key}, path: {self.module_path}"
+            )
+
+        run_method = getattr(mod, self.kernel_name).run
+
+        return functools.partial(
+            run_method,
+            *input_tensors,
+            output_tensor,
+            *self.extra_args,
+            grid=self.grid,
+            num_stages=self.num_stages,
+            num_warps=self.num_warps,
+            stream=torch.cuda.current_stream().cuda_stream,
+        )
+
+    def __str__(self) -> str:
+        return f"{self.kernel_name=}, {self.module_path=}, {self.module_cache_key=}"
+
+
+class CUDABenchmarkRequest(BenchmarkRequest):
+    def __init__(
+        self,
+        kernel_name: str,
+        input_tensor_meta: List[TensorMeta],
+        output_tensor_meta: TensorMeta,
+        extra_args: Dict[str, Any],
+        source_code: str,
+    ):
+        super().__init__(kernel_name, input_tensor_meta, output_tensor_meta, extra_args)
+        self.source_code = source_code
+        self.workspace_size: int = 0
+        self.workspace: Optional[Tensor] = None
+        self.DLL: Optional[DLLWrapper] = None
+        self.hash_key: str = ""
+        self.source_file: str = ""
+        self.hash_key, self.source_file = CUDACodeCache.write_source_code(
+            self.source_code, "so"
+        )
+
+    def make_run_fn(
+        self, *input_tensors: torch.Tensor, output_tensor: torch.Tensor
+    ) -> Callable[[], None]:
+        self.DLL, self.hash_key, self.source_file = CUDACodeCache.load(
+            self.source_code, "so"
+        )
+        args = [
+            c_void_p(tensor.data_ptr())
+            for tensor in list(input_tensors) + [output_tensor]
+        ]
+        log.debug(
+            "make_run_fn: "
+            f"{self.kernel_name=}, {self.source_file=}, {self.hash_key=}, {self.DLL=}"
+            f"{args=}, {self.extra_args=}"
+        )
+        run_method = getattr(self.DLL, self.kernel_name)
+        stream_ptr = c_void_p(torch.cuda.current_stream().cuda_stream)
+
+        # Retrieve workspace_size and initialize workspace.
+        c_workspace_size = c_size_t()
+        run_method(
+            *args,  # input ptrs and output ptrs
+            *self.extra_args,
+            byref(
+                c_workspace_size
+            ),  # set workspace size ptr to retrieve workspace size
+            None,  # null workspace ptr
+            stream_ptr,
+        )
+        self.workspace_size = c_workspace_size.value
+        assert (
+            self.workspace_size == 0
+        ), "Autotune cache needs to be updated to support non-zero workspace_size!"
+        self.workspace = torch.empty(
+            self.workspace_size, dtype=torch.uint8, device="cuda"
+        )
+
+        # Generate partial function.
+        return functools.partial(
```

Comment:
Not saying this is a bad idea here to use functools.partial, but it can make debugging harder, as functools.partial binds it into a native object, leaving no trace in callstacks where the code is coming from. So when the function returned by functools.partial is called at a different callsite later on, it is hard to figure out which function is being called, without actually using a debugger or tracing back the origin of the function. On the other hand, it's likely slightly faster than using a python wrapper function.